### PR TITLE
Updating a "how to contribute" section

### PR DESCRIPTION
**Pull request by rvs**
See here for how it actually looks like https://github.com/rvs/gpdb#contributing

This is very much available for peer review and comments (especially the peer review and comments section ;-))

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -282,8 +282,151 @@ Known issues:
...
+We recommend using ```git diff --color``` when reviewing your changes so that you
+don't have any spurious whitespace issues in the code that you submit.
+
+All new functionality that is contributed to Greenplum should be covered by regression
+tests that are contributed alongside it.
+
+### Testing guidelines
+
+At the very minimum you should always be running 
+```make installcheck-good```
```
Comment (ashwinstar) [inline README.md]: `make installcheck-world`,  installcheck-good was in olden days.

File: README.md [hunk 2]
```diff
@@ -282,8 +282,151 @@ Known issues:
...
+that we keep reducing the delta between the two projects, but also so that any change
+that is relevant to PostgreSQL can benefit from a much broader review of the upstream
+PostgreSQL community. In general it is a good idea to keep both code bases handy so
+you can be sure whether your changes may need to be forward ported.
+
+### Submission timing
+
+To improve the odds of the right discussion of your patch or idea happening, pay attention 
+to what the community work cycle is. If for example you send in a brand new idea in the 
+beta phase, don't be surprised if no one is paying attention because they are focused on
```
Comment (ashwinstar) [inline README.md]: s/they/we

File: README.md [hunk 3]
```diff
@@ -282,8 +282,151 @@ Known issues:
...
+recommend following a stream of pull reviews to anybody who's interested in becoming
+a long term contributor to Greenplum. As [Linus would say](https://en.wikipedia.org/wiki/Linus's_Law) 
+"given enough eyeballs, all bugs are shallow".
+
+One outcome of the peer review could be a consensus that you need to modify your
+pull request in certain ways. GitHub allows you to push additional commits into
+a branch from which a pull request was sent. Those additional commits will be then
+visible to all of the reviewers.
+
+Note that Greenplum
```
Comment (ashwinstar) [inline README.md]: Not clean what this meant, incomplete sentence ?

Comment (rvs) [inline README.md]: Fixed

File: README.md (hunk 2 above)
Comment (rvs) [inline README.md]: Good catch.

File: README.md (hunk 1 above)
Comment (rvs) [inline README.md]: Aha! Thanks for clarifying.

File: README.md [hunk 4]
```diff
@@ -282,8 +282,149 @@ Known issues:
...
+you can be sure whether your changes may need to be forward ported.
+
+### Submission timing
+
+To improve the odds of the right discussion of your patch or idea happening, pay attention 
+to what the community work cycle is. If for example you send in a brand new idea in the 
+beta phase, don't be surprised if no one is paying attention because we are focused on 
+release work. Come back when the beta is done, please!
+
+You can read more on Greenplum release policy and timing in the RELEASE.md
```
Comment (cjcjameson) [inline README.md]: There is no such thing at the moment.

@tom-meyer @pivotal-mike can we rewrite or delete https://github.com/greenplum-db/gpdb/blob/master/src/tools/RELEASE_CHANGES in a separate PR?

**Comment by rvs** [comment]
Thanks @cjcjameson !

File: README.md [hunk 5]
```diff
@@ -282,8 +282,148 @@ Known issues:
 
 ## Contributing
 
-If you have not previously done so, please fill out and
-submit the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum).
+Greenplum is maintained by a core team of developers with commit rights to the
+[main gpdb repository](https://github.com/greenplum-db/gpdb) on GitHub. At the
+same time, we are very eager to receive contributions from anybody in the wider
+Greenplum community. This section covers all you need to know if you want to see
+your code or documentation changes be added to Greenplum repository and to the
```
Comment (danielgustafsson) [inline README.md]: Just my $0.02, but I would phrase it as "..added to Greenplum."

File: README.md [hunk 6]
```diff
@@ -282,8 +282,148 @@ Known issues:
...
+the reviewer's job much easier and increases the timeliness of feedback.
+
+When it comes to C and C++ parts of Greenplum, we try to follow 
+[PostgreSQL Coding Conventions](https://www.postgresql.org/docs/devel/static/source.html).
+In addition to that we require that:
+   * All Python code passes [Pylint](https://www.pylint.org/)
+   * All Go code is formatted according to [gofmt](https://golang.org/cmd/gofmt/)
+
+We recommend using ```git diff --color``` when reviewing your changes so that you
+don't have any spurious whitespace issues in the code that you submit.
```
Comment (danielgustafsson) [inline README.md]: Isn't it easier to recommend `git diff --check` since it will actually say exactly what and where the error is?

File: README.md [hunk 7]
```diff
@@ -282,8 +282,148 @@ Known issues:
...
+
+All new functionality that is contributed to Greenplum should be covered by regression
+tests that are contributed alongside it.
```
Comment (danielgustafsson) [inline README.md]: I agree, and I at the same time want to lower the barrier to entry, sort of along the lines of: "..regression tests, and documentation, that are contributed alongside it. If you are uncertain on how to test, or document your work, please raise the question on the gpdb-dev mailinglist and the developer community will do its best to help you."

File: README.md [hunk 8]
```diff
@@ -282,8 +282,148 @@ Known issues:
...
+To improve the odds of the right discussion of your patch or idea happening, pay attention 
+to what the community work cycle is. For example, if you send in a brand new idea in the 
+beta phase, don't be surprised if no one is paying attention because we are focused on 
+release work. Come back when the beta is done, please!
+
+You can read more on Greenplum release policy and timing in the RELEASE.md
+
+### Patch submission
+
+Once you are ready to share your work with the Greenplum core team and the rest of
```
Comment (danielgustafsson) [inline README.md]: "core team" sounds like the PostgreSQL counterpart. Given that most contributors will be familiar with postgres it's probably best to avoid nomenclature that can confuse. How about "Greenplum committers"?

File: README.md [hunk 9]
```diff
@@ -282,8 +282,148 @@ Known issues:
...
+ask on the developer's mailing list, but make sure to include a direct link
+to a pull request in your email.
+
+### Patch review
+
+A submitted pull request with passing validation checks is assumed to be available 
+for peer review. Peer review is the process that ensures that contributions to Greenplum 
+are of high quality and align well with the road map and community expectations. Every
+member of the Greenplum community is encouraged to review pull requests and provide
+feedback. Since you don't have to be a core team member to be able to do that, we
```
Comment (danielgustafsson) [inline README.md]: s/core team member/committer/ ? See above comment.

File: README.md (hunk 7 above)
Comment (Eulerizeit) [inline README.md]: +1 on this. We should be committing code, tests AND documentation. All of them work together in helping review the code. 

@schubert @pivotal-mike @professor I assume this is the case with people who are already committers as well? It was discussed but I'm not sure if it was decided on.

File: README.md [hunk 10]
```diff
@@ -282,8 +282,148 @@ Known issues:
...
+first and sharing your proposal with us before you spend a lot of time writing
+code. Even when your proposal gets validated by the community, we still recommend
+doing the actual work as a series of small, self-contained commits. This makes
+the reviewer's job much easier and increases the timeliness of feedback.
+
+When it comes to C and C++ parts of Greenplum, we try to follow 
+[PostgreSQL Coding Conventions](https://www.postgresql.org/docs/devel/static/source.html).
+In addition to that we require that:
+   * All Python code passes [Pylint](https://www.pylint.org/)
+   * All Go code is formatted according to [gofmt](https://golang.org/cmd/gofmt/)
```
Comment (Eulerizeit) [inline README.md]: Should we add the Postgres Linter here as well?  Should we work on the postgres Linter so it doesn't have the issues @danielgustafsson and @hlinnaka pointed out?

Comment (danielgustafsson) [inline README.md]: I don't find it helpful (or even interesting) to impose pgindent on submissions, it just raises the barrier to entry.

(Also, until all our code passes pgindent we can't since it would generate enormous diffs with lots unrelated changes)

**Comment by schubert** [comment]
+1 okay with all of this for now. We can revisit tweaks as questions arise or new developers read through it and give feedback.

File: README.md (hunk 6 above)
Comment (rvs) [inline README.md]: I was actually following Postgres advice here. I don't mind your suggestion at all -- but you may want to update Postgres contributions doc as well then.

File: README.md (hunk 8 above)
Comment (rvs) [inline README.md]: It was sort of done on purpose to signal similarities in the governance model. If you feel strongly about the confusing nature of this -- please let me know (best via a PR ;-))
